### DISCUSSION TITLE
Missing steps in documentation

### DISCUSSION THREAD

**Discussion by GullyTechIO** [llm-tools/embedJs#14] [General]
Hello,

I am following the instructions line by line.

1. npm i @llm-tools/embedjs

2. Pick an LLM
`const ragApplication = await new RAGApplicationBuilder()
.setModel(SIMPLE_MODELS.OPENAI_GPT3_TURBO)`
3. Pick a vector database - ok, how about lanceDB
> npm install vectordb
```
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @langchain/community@0.0.35
npm WARN Found: vectordb@0.4.16
npm WARN node_modules/vectordb
npm WARN   vectordb@"*" from the root project
npm WARN   1 more (@llm-tools/embedjs)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peerOptional vectordb@"^0.1.4" from @langchain/community@0.0.35
npm WARN node_modules/@langchain/community
npm WARN   @langchain/community@"^0.0.35" from @llm-tools/embedjs@0.0.66
npm WARN   node_modules/@llm-tools/embedjs
npm WARN
npm WARN Conflicting peer dependency: vectordb@0.1.19
npm WARN node_modules/vectordb
npm WARN   peerOptional vectordb@"^0.1.4" from @langchain/community@0.0.35
npm WARN   node_modules/@langchain/community
npm WARN     @langchain/community@"^0.0.35" from @llm-tools/embedjs@0.0.66
npm WARN     node_modules/@llm-tools/embedjs
```
wow I hope none of that is important

`.setVectorDb(new LanceDb({ path: path.resolve('/db') }))`
4. Add a loader
`.addLoader(new WebLoader({ url: 'https://en.wikipedia.org/wiki/Formula_One' }))`

"That's it!"

Cool!

> node rag.js

```
const ragApplication = await new RAGApplicationBuilder()
                       ^^^^^
SyntaxError: await is only valid in async functions and the top level bodies of modules
    at internalCompileFunction (node:internal/vm:77:18)
    at wrapSafe (node:internal/modules/cjs/loader:1288:20)
    at Module._compile (node:internal/modules/cjs/loader:1340:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1435:10)
    at Module.load (node:internal/modules/cjs/loader:1207:32)
    at Module._load (node:internal/modules/cjs/loader:1023:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12)
    at node:internal/main/run_main_module:28:49
```
Alright, removed all of the "await"s because this is being run directly.

```
const ragApplication = new RAGApplicationBuilder()
                       ^
ReferenceError: RAGApplicationBuilder is not defined
```

Here's the first real snag. Documentation doesn't mention this, but Node needs you to import all of the objects you're likely to use. So I try this:
`const { RAGApplicationBuilder, WebLoader, YoutubeLoader, SitemapLoader, SIMPLE_MODELS } = require('@llm-tools/embedjs');`

```
.setVectorDb(new LanceDb({ path: path.resolve('/db') }))
                 ^
ReferenceError: LanceDb is not defined
```

Now I need to figure out what CONST function can be used to import LanceDB. So I eventually find this:

`const { LanceDb } = require('vectordb');`

New error: 

```
.setVectorDb(new LanceDb({ path: path.resolve('/db') }))
                                 ^
ReferenceError: path is not defined
```

What does it mean PATH is not defined? ok, so I add another CONST "const { path } = require('path');"

```
.setVectorDb(new LanceDb({ path: path.resolve('/db') }))
             ^
TypeError: LanceDb is not a constructor
```

Not a constructor?? What is it then? Right, time to look at your example scripts to see how you handle it.

Nothing. No examples for lancedb. Let's do a Google search for "setVectorDb(new LanceDb"

> 4 results. Two of them link right back here. The others are useless.

Conclusion: This project greatly oversells how easy it is to use.

**Comment by GullyTechIO** [comment]
And of course not a single google result for "LanceDb is not a constructor". Does anyone actually use this project or is it a thought experiment?

**Comment by GullyTechIO** [comment]
Maybe I'll try HNSWLib instead.

1. Install HNSWLib package in your project
> npm install hnswlib-node

```
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @langchain/community@0.0.35
npm WARN Found: hnswlib-node@3.0.0
npm WARN node_modules/hnswlib-node
npm WARN   hnswlib-node@"*" from the root project
npm WARN
npm WARN Could not resolve dependency:
npm WARN peerOptional hnswlib-node@"^1.4.2" from @langchain/community@0.0.35
npm WARN node_modules/@langchain/community
npm WARN   @langchain/community@"^0.0.35" from @llm-tools/embedjs@0.0.66
npm WARN   node_modules/@llm-tools/embedjs
npm WARN
npm WARN Conflicting peer dependency: hnswlib-node@1.4.2
npm WARN node_modules/hnswlib-node
npm WARN   peerOptional hnswlib-node@"^1.4.2" from @langchain/community@0.0.35
npm WARN   node_modules/@langchain/community
npm WARN     @langchain/community@"^0.0.35" from @llm-tools/embedjs@0.0.66
npm WARN     node_modules/@llm-tools/embedjs
npm WARN ERESOLVE overriding peer dependency
npm WARN While resolving: @langchain/community@0.0.35
npm WARN Found: vectordb@0.4.16
npm WARN node_modules/vectordb
npm WARN   peerOptional vectordb@"^0.4.11" from @llm-tools/embedjs@0.0.66
npm WARN   node_modules/@llm-tools/embedjs
npm WARN     @llm-tools/embedjs@"^0.0.66" from the root project
npm WARN   1 more (the root project)
npm WARN
npm WARN Could not resolve dependency:
npm WARN peerOptional vectordb@"^0.1.4" from @langchain/community@0.0.35
npm WARN node_modules/@langchain/community
npm WARN   @langchain/community@"^0.0.35" from @llm-tools/embedjs@0.0.66
npm WARN   node_modules/@llm-tools/embedjs
npm WARN
npm WARN Conflicting peer dependency: vectordb@0.1.19
npm WARN node_modules/vectordb
npm WARN   peerOptional vectordb@"^0.1.4" from @langchain/community@0.0.35
npm WARN   node_modules/@langchain/community
npm WARN     @langchain/community@"^0.0.35" from @llm-tools/embedjs@0.0.66
npm WARN     node_modules/@llm-tools/embedjs
npm ERR! code ERESOLVE
npm ERR! ERESOLVE could not resolve
npm ERR!
npm ERR! While resolving: @llm-tools/embedjs@0.0.66
npm ERR! Found: hnswlib-node@3.0.0
npm ERR! node_modules/hnswlib-node
npm ERR!   hnswlib-node@"*" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peerOptional hnswlib-node@"^2.1.1" from @llm-tools/embedjs@0.0.66
npm ERR! node_modules/@llm-tools/embedjs
npm ERR!   @llm-tools/embedjs@"^0.0.66" from the root project
npm ERR!
npm ERR! Conflicting peer dependency: hnswlib-node@2.1.1
npm ERR! node_modules/hnswlib-node
npm ERR!   peerOptional hnswlib-node@"^2.1.1" from @llm-tools/embedjs@0.0.66
npm ERR!   node_modules/@llm-tools/embedjs
npm ERR!     @llm-tools/embedjs@"^0.0.66" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR!
npm ERR!
npm ERR! For a full report see:
npm ERR! C:\Users\****\AppData\Local\npm-cache\_logs\2024-04-05T04_56_33_926Z-eresolve-report.txt

npm ERR! A complete log of this run can be found in: C:\Users\****\AppData\Local\npm-cache\_logs\2024-04-05T04_56_33_926Z-debug-0.log
```

**Comment by GullyTechIO** [comment]
It looks like @langchain/community is compatible with hnswlib-node@1.4.2
but @llm-tools/embedjs would like hnswlib-node@2.1.1

**Comment by GullyTechIO** [comment]
And package "@langchain/community" requires vectordb version 0.1.4 as a peer dependency, while another package "@llm-tools/embedjs" requires vectordb version 0.1.19.

**Comment by adhityan** [comment]
Hello @GullyTechIO I just responded to your issue and I will try and answer your questions here as well. Here's a few things for benefit - 

1. The library is meant to be used by people familiar with NodeJs. The documentation does not explicitly mention that importing dependencies is necessary in NodeJs. It also does not provide examples on how to use built in node packages like `path`. Since your acount was created just to post this disucssion and create the issue, it is unclear to me if you are familiar with NodeJs already. I do not plan to expand the documentation to start explaining NodeJs fundamentals as it would be counterproductive and not relevant to the intended audience of this library.

2. The library references to the latest (at the time of last build) dependencies of various packages. Wherever other versions of these dependencies are referenced in other packages, you will need to add an `overrides` section in your package.json. The library comes with it it's own overrides section to prevent any errors out of the box. The slack-bot example repo uses LanceDB and works just well. These warnings (visible in versbose mode) are useful to understand how npm resolves depenedencies but they are not errors.

<img width="401" alt="Screenshot 2024-04-05 at 10 57 51" src="https://github.com/llm-tools/embedJs/assets/5626502/9c7a8a74-4f8c-4d58-a159-bddc09386975">

4. For HNSW-lib, you tried installing the latest version 3.0.0 which was published after the last release of the package and was not supported by this library at the time of your review. The NPM error clearly spells out which version you had to install and following that, you would have rid yourself of the error. Since then, I have added support for v3.0.0 of HNSW-lib, so you can also install the latest version of HNSW-lib if you want to. 

Note, if you need a version that an open source library does not support already, consider sending in a PR as opposed to demanding urgent updates. This is voluntary unpaid work. Again, since you created a Github account just to critique this library, I am not sure you are familair with opensource already - but, we appreciate people being polite and contributing back.

If you really need help installing and getting things working, drop me a note and we can have a call where I can help you get it working. I feel you may be struggling a little bit because of your unfamiliariy with NodeJs but I am happy to help.

**Reply by GullyTechIO** [reply]
Thank you for your expedient response. I want to apologise for the false sense of urgency, and especially apologise for my harsh words above. I am embarrassed by them. I got too excited at the prospect of an easy RAG and lashed out in impotent frustration when the instructions did not work.

I have minor node experience, but only in contexts where I could find or figure out the name of the dependencies. I'll knuckle down and give it the proper attention it deserves to figure out why HNSWDb is not a constructor.

**Comment by GullyTechIO** [comment]
I've installed hnswlib-node
Added this line to my code with no errors
`const { HNSWDb } = require('hnswlib-node');`

Then added this to my ragApplication
`.setVectorDb(new HNSWDb())`

However
`TypeError: HNSWDb is not a constructor`

I've noticed that the hnswlib-node project does not mention HNSWDb. 
HNSWDb is created by _this_ project
https://github.com/llm-tools/embedJs/blob/37c28761b1f47f20fd9c285a7b75de3dc9d016b6/src/vectorDb/hnswlib-db.ts#L7

So am I wrong to `const { HNSWDb } = require('hnswlib-node');` when the constructor is in the code of embedJs?

Whether I do that, or add HNSWDb to the list of objects required from @llm-tools/embedjs, I am still given the same error `TypeError: HNSWDb is not a constructor`

**Comment by adhityan** [comment]
Here's how you should go about it -

1. Install hnswlib-node ^3.0.0 
`npm i hnswlib-node` should do the trick. You should see no errors or warnings (in non verbose mode).

2. Import HNSWDb 
`import { HNSWDb } from '@llm-tools/embedjs/vectorDb/hnswlib'`
For commonJs, the require syntax should work. Make sure you are fetching it from the exported path `@llm-tools/embedjs/vectorDb/hnswlib`

3. Set HNSWDb as your choice of vector DB
`.setVectorDb(new HNSWDb())`

I tested this with the example slack bot using Node v20.10.0 just now and everything works fine. The only difference between my setup and yours is I use the import syntax but I don't think that should be an issue as the library is compiled for both formats.

**Reply by adhityan** [reply]
Ideally your IDE should auto-complete the import statement and path but you can use the above path manually in case your IDE does not support node auto-completes..

**Comment by adhityan** [comment]
You should be only using exports from embedJs everywhere. No objects from `hnswlib-node` need to be used directly. The reason you need to separately install vector db libraries is that - I do not want to package all dependencies for all vector databases in the library. So people can pick and use the vector db they want and only install the relevant packages.

**Comment by GullyTechIO** [comment]
Thanks very much, this is where I am now:

```
const { RAGApplicationBuilder, WebLoader, SIMPLE_MODELS } = require('@llm-tools/embedjs');
const { LanceDb } = require('@llm-tools/embedjs/vectorDb/lance');
require('dotenv').config();

const main = async () => {
    const ragApplication = new RAGApplicationBuilder()
        .setModel(SIMPLE_MODELS.OPENAI_GPT3_TURBO)
        .setVectorDb(new LanceDb({ path: '.db' }))
        .addLoader(new WebLoader({ url: 'https://en.wikipedia.org/wiki/Formula_One' }))
        .build();
    console.log(ragApplication.query('Please provide a one paragraph summary of Formula One'));
};

main();
```

`TypeError: ragApplication.query is not a function`

I tried with and without putting "await" before ragApplication.query and I can see that query definitely is in 
https://github.com/llm-tools/embedJs/blob/37c28761b1f47f20fd9c285a7b75de3dc9d016b6/src/core/rag-application.ts#L195

Is there something else I have to add to the first line to let it recognise the functions of RagApplication objects created by RagApplicationBuilder?

I tried

`const { RAGApplicationBuilder, WebLoader, SIMPLE_MODELS, RAGApplication } = require('@llm-tools/embedjs');`

based on 
https://github.com/llm-tools/embedJs/blob/37c28761b1f47f20fd9c285a7b75de3dc9d016b6/src/core/rag-application.ts#L14

but no change.

**Comment by GullyTechIO** [comment]
Update!

I ditched my efforts and copied parts of commands from examples/simple/src/index.ts

```
const { RAGApplicationBuilder, WebLoader} = require('@llm-tools/embedjs');
const { LanceDb } = require('@llm-tools/embedjs/vectorDb/lance');
require('dotenv').config();

const main = async () => {
	const llmApplication = await new RAGApplicationBuilder().setVectorDb(new LanceDb({ path: '.db' })).build();
	llmApplication.addLoader(new WebLoader({ url: 'https://old.reddit.com/r/news/' }));
	console.log((await llmApplication.query('Provide a quick summary of items on https://old.reddit.com/r/news/')).result);
};

main();
```

It correctly picked out 10 items from that page and rephrased the headlines.

🥳

**Comment by adhityan** [comment]
This is great to hear. Anything I can do to help more?
